pkg/sql/schemachanger/scbuild/build.go

```diff
@@ -104,10 +104,16 @@ func Build(
 	current := make([]scpb.Status, 0, len(bs.output))
 	version := dependencies.ClusterSettings().Version.ActiveVersion(ctx)
 	withLogEvent := make([]scpb.Target, 0, len(bs.output))
+	var extraTargets []struct {
+		e elementState
+		t scpb.Target
+	}
 	for _, e := range bs.output {
-		if e.metadata.Size() == 0 {
+		if e.metadata.Size() == 0 && !shouldElementBeRetainedWithoutMetadata(e.element) {
```

Comment:
Please help me understand the overall purpose of including those `TableData` element and `IndexData` element in the TargetState even if they are not added/dropped in the builder. In other words, how exactly do they help "hinting" if a backfill is needed

Reply:
They allow us to know if a table is empty, which allows us to skip backfill steps completely. The other alternative is to use the DESCRIPTOR_ADDED state, but that breaks if we enter a more transactional world. Since you should be able to insert on those

Reply:
Updated the comments here:
			// Exceptions are TableData/IndexData elements which allow our planning
			// execution to skip certain transitions and fences like the two version
			// invariant. We will only keep these for descriptors going through
			// a transition.